### DISCUSSION TITLE
Custom translate source

### DISCUSSION THREAD

**Discussion by chunibyocola** [chunibyocola/sc-translator-crx#31] [Show and tell]
# Overview
“Custom translate source" is a new feature of ScTranslator. You can add custom sources which built by yourself or others.
# Are there any requirements for adding custom source?
Yes, the source you want to add must response correct format that extension required to every requests:
```typescript
type DataFromCustomSource = {
    result: string[]; // The result of the translate. Must be an array of string.
    from: string; // The source language of the text.
    to: string; // The target language.
    dict?: string[]; // Word explanation. Assign nothing if there is no explanation.
    phonetic?: string; // The phonetic of the word. Assign nothing if there is no phonetic.
    related?: string[]; // Related words like "prefer" to "preferred". Assign nothing if there is no related words.
    example?: string[]; // Examples of the word. Assign nothing if there is no examples.
} | {
    code: string; // Error code from server if there are some issues happened.
};
```
After you clicking the "Add" icon, extension will send a request to the url you inputed with body like:
```typescript
{
    text: 'test',
    from: 'auto',
    to: 'en',
    userLang: 'en',
    preferred: ['en', 'ja']
}
```
If response meet the requirements, the custom source can be successfully added.
# How to build your custom source?
It is easy to build, but it requires you have a certain backend programming knowledge.

I will explain it will [Node.js](https://nodejs.org/en/).
## Initialize project
Environment (The environment that I develop this example):
```
OS: Windows 10
Yarn: 1.22.7
Node: v16.13.0
```
Scripts:
```
mkdir server-example
cd server-example
yarn
yarn add express body-parser
```
Create a file named `index.js` in folder `server-example`.

Add `"main": "index.js"` and `"type": "module"` in `package.json`:
```
{
  "main": "index.js",
  "type": "module",
  "dependencies": { ... }
}
```
Copy the codes in `index.js`:
```javascript
import express from 'express';
import bodyParser from 'body-parser';

const app = express();
const port = 3000;

app.use(bodyParser.json());

// The language codes map that your server supports.
const supportedLangCodes = {
    'auto': 'auto',
    'en': 'English',
    'ja': 'Japanese',
    'zh-CN': 'Chinese'
};

// Write your "detect" codes here or ignore it.
const detectLanguage = async (text) => {
    return undefined;
};

const translateSource1 = async (params) => {
    // params: { text: 'hello', from: 'en', to: 'zh-CN', userLang: 'zh-CN' }
    // mock data here
    return {
        result: ['你好'],
        from: 'en',
        to: 'zh-CN',
        dict: ['感叹词: 你好!, 喂!'],
        phonetic: '[həˈlō]'
    };
};

const translateSource2 = async (params) => {
    // params: { text: 'hello', from: 'en', to: 'ja', userLang: 'ja' }
    // mock data here
    return {
        result: ['こんにちは'],
        from: 'en',
        to: 'ja',
        dict: ['名词: 今日は'],
        phonetic: '[həˈlō]'
    };
};

// Should only work when from is "auto".
const dealWithFromAndTo = async (text, from, to, preferred) => {
    if (from === 'auto') {
        const detectedCode = await detectLanguage(text);

        if (detectedCode) {
            from = detectedCode;

            if (to === from) {
                const [preferredLanguage, secondPreferredLanguage] = preferred;

                to = (from === preferredLanguage ? secondPreferredLanguage : preferredLanguage);
            }
        }
    }

    return [from, to];
};

app.post('/translateSource1', async (req, res) => {
    let { text, from, to, userLang, preferred } = req.body;

    [from, to] = await dealWithFromAndTo(text, from, to, preferred);

    if (!supportedLangCodes[from] || !supportedLangCodes[to]) {
        res.send({ code: 'LANGUAGE_NOT_SOPPORTED' });
        return;
    }

    const data = await translateSource1({ text, from, to, userLang });
    res.send(data);

    return;
});

app.post('/translateSource2', async (req, res) => {
    let { text, from, to, userLang, preferred } = req.body;
    
    [from, to] = await dealWithFromAndTo(text, from, to, preferred);

    if (!supportedLangCodes[from] || !supportedLangCodes[to]) {
        res.send({ code: 'LANGUAGE_NOT_SOPPORTED' });
        return;
    }

    const data = await translateSource2({ text, from, to, userLang });
    res.send(data);

    return;
});

app.listen(port, () => {
    console.log(`Example app listening at http://localhost:${port}`);
});
```
Run this project with `node index.js`.
### supportedLangCodes
`supportedLangCodes` is a map to check the language that server support or not from client's request.
The language code from extension could only be one of the [language codes map](https://github.com/chunibyocola/sc-translator-crx/blob/master/src/public/translate/google/lang-code.ts).
### detectLanguage
`detectLanguage` is a function to detect what language the text is.
If the translate service you use is already provide such a funtion, you can ignore it. Or you should ensure there is a way to detect the language.
### dealWithFromAndTo
`dealWithFromAndTo` is a funtion to detect(or not) the language of text and decide the target language depend on the source language.
### Request body from client
You can see there is deconstruct syntax in `app.post`:
```
let { text, from, to, userLang, preferred } = req.body;
```
The request body from client will follow the type:
```typescript
type FetchCustomSourceJSON = {
    text: string; // The text that user want to translate.
    from: string; // The source language of the text. (Often "auto")
    to: string; // The target language. (Not possible to be "auto", will be preferred language if source language is "auto")
    userLang: string; // User's language. (Come from "window.navigator.language" in JS)
    preferred: [string, string]; // User's preferred language and second preferred language.
};
```
### Errors
If there are errors happened in server or you want to tell user something, you can send [fixed strings](https://github.com/chunibyocola/sc-translator-crx/blob/master/src/public/translate/error-codes.ts) to client like:
```javascript
res.send({ code: 'LANGUAGE_NOT_SOPPORTED' });
return;
```
or
```javascript
res.send({ code: 'NO_RESULT' });
return;
```
, etc.

You can also send sentence like:
```javascript
res.send({ code: 'The things I want to tell.' });
return;
```
**Don't assign to `code` with anything if there are no errors!**
### Multiple sources
You can see there are two translate sources here, `translateSource1` and `translateSource2`.

Yes, you can provide mutiple sources in a single server.

By adding URL in custom translate source in options page two times: `http:127.0.0.1:3000/translateSource1` and `http:127.0.0.1:3000/translateSource2`.
# Feel free to leave a comment down bellow!

**Comment by oott123** [comment]
是否有机会支持整页翻译的自定义接口呢？

**Reply by chunibyocola** [reply]
现在的网页翻译是会考虑到 DOM 的嵌套关系才能做到仅显示译文也能保持原来的链接和句子结构，这归功于专用的网页翻译源。如果使用自定义的翻译源的话，就不能选择仅显示译文了。考虑到这一点，我不会太过于倾向支持这个功能，但我也知道很多人都想用其他一些翻译源来翻译网页，比如 DeepL。需要考虑一下是否有这个必要。

**Reply by oott123** [reply]
@chunibyocola 大概可以要求自定义的服务器也实现同样的接口？

**Reply by chunibyocola** [reply]
这不现实，因为现有的非网页翻译专用的源都做不到。而面向网页翻译的源也已经自带了，目前我觉得没有能比谷歌网页翻译更好的网页翻译源了。

**Reply by oott123** [reply]
@chunibyocola  一般提供 API 的翻译源都会有 XML 标记或者 HTML 标记处理的。DeepL 就有这个。
* https://www.deepl.com/docs-api/handling-html-(beta)/
* https://www.deepl.com/zh/docs-api/handling-xml/

**Reply by chunibyocola** [reply]
我想到了一个方法是
```HTML
<div>This <span>is</span> a sentence.</div>
<div>This <span>is another sentence.</span></div>
```
转化为
```JavaScript
[
    ["This ", "is", " a sentence."]
    ["This ", "is another sentence."]
]
```
后再交给服务器自行拼接，返回格式一定要为
```JavaScript
[
    [string * 3]
    [string * 2]
]
```
的数组。

这样应该可以。

**Comment by mrsimpson** [comment]
This is really amazing and to me the key feature of sc-translator, as it enables our corporate translator tool to be integrated easily. Without needing to fork the extension! 

Just wanted to take the time to leave a huge ❤️  for this (and to get you into a good mood before I open up new feature requests ;))
